DeploymentAgentReconfigurationScripts/TFS2018 Update2/HardwareMoveRemoveOldAgent.ps1

```diff
@@ -0,0 +1,178 @@
+
+Output:
+If action is set to apply, this script will
+1. Delete the existing unusable offline agent and re-configure a new agent in with same name and properties including tags.
+2. All existing deployment history will be deleted. 
+
+Example usage:
+.\HardwareMoveRemoveOldAgent.ps1 -targetTFSUrl <newTfsUrl> -patToken <PAT-for-new-TFS>
+.\HardwareMoveRemoveOldAgent.ps1 -targetTFSUrl <newTfsUrl> -patToken <PAT-for-new-TFS> -existingAgentFolder <AgentFoilder (C:\vstsagents\A1)>
+.\HardwareMoveRemoveOldAgent.ps1 -targetTFSUrl <newTfsUrl> -patToken <PAT-for-new-TFS> -existingAgentFolder <AgentFoilder (C:\vstsagents\A1)> -action 'apply'
+#>
+
+param([string]$targetTFSUrl,
+      [string]$patToken,
+      [string]$existingAgentFolder = "",
+      [string]$agentDownloadUrl = 'https://vstsagentpackage.azureedge.net/agent/2.131.0/vsts-agent-win-x64-2.131.0.zip',
+      [string]$action = "PrintEffect")
+
+$ErrorActionPreference="Stop"
+
+# Basic input validations
+If (-NOT ([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent() ).IsInRole( [Security.Principal.WindowsBuiltInRole] �Administrator�)){ 
+    throw "Run command in an administrator PowerShell prompt"
+};
+
+If ($PSVersionTable.PSVersion -lt (New-Object System.Version("3.0"))){
+    throw "The minimum version of Windows PowerShell that is required by the script (3.0) does not match the currently running version of Windows PowerShell."
+}
+
+
+# Auto detect existing agent folder
+if ($existingAgentFolder -eq ""){
+    $ServiceNamePrefix = 'vstsagent*'
+    $agentServices = Get-Service -Name $ServiceNamePrefix
+    if ($agentServices.Count -eq 1){
+        $serviceName = $agentServices[0].Name;
+        $service = get-wmiobject -query "select * from win32_service where name = `'$serviceName`'";
+        $serviceExePath = $service.pathname;
+        # $serviceExePath will have value like "C:\vstsAgent2\A8\bin\AgentService.exe"
+        $existingAgentFolder = $serviceExePath.Substring(1, $serviceExePath.Length - 22);
+    }
+}
+
+if ($existingAgentFolder -eq ""){
+    throw "Not able to auto detect existing agent folder. Provide the existingAgentFolder as input parameter. You can find it in agent capabilities UI as Agent.HomeDirectory.  Generally it is in C:\vstsagent folder.";
+}
+
+cd $existingAgentFolder;
+
+if (-not (Test-Path '.\.agent')){
+    throw "No agent installed in this path. Please run this script from Agent Home Directory. Generally it is in C:\vstsagent folder."
+}
+
+
+# Collect information about the existing agent
+$sourceTFSUrl = "NoAbleToReadAgentFile"
+$collectionName = "NoAbleToReadAgentFile"
+$projectName = "NoAbleToReadAgentFile"
+$deploymentGroupId = "NoAbleToReadAgentFile"
+$deploymentGroupName = "NoAbleToReadAgentFile"
+$agentName = "NoAbleToReadAgentFile"
+$tags = "NoAbleToReadAgentFile"
+$deploymentMachineId = "NoAbleToReadAgentFile"
+
+foreach($line in Get-Content .\.agent) {
```

Comment:
this loop is very tricky, can we find any better way where like creating object out of json, at the end we write .agent file from on custom C# object, some reverse engineering

Reply:
$t = (Get-Content "D:\Agents\1\.agent") -join "`n" | ConvertFrom-Json

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/convertfrom-json?view=powershell-6

Reply:
$t.projectId

Reply:
we can try that. I will check